### PR TITLE
Add support for instructions using more registers in exegesis annotator

### LINKED ISSUE #251: Exegesis annotator failing in machine code verification
The exegesis annotator is currently failing on at least one block with the following machine code verification error.

```
*** Bad machine code: Using an undefined physical register ***
 2024-10-29 23:53:48.280841-07000   - function:    foo
 2024-10-29 23:53:48.280848-07000   - basic block: %bb.0  (0x7fe694a284c8)
 2024-10-29 23:53:48.280959-07000   - instruction: MOVSQ $rdi, $rsi, $noreg, implicit-def $edi, implicit-def $esi, implicit $edi, implicit $esi, implicit $df
 2024-10-29 23:53:48.281092-07000   - operand 7:   implicit $df
```

It looks like this needs to get fixed during the annotation stage.

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch adds explicit support in the exegesis annotator for instructions that use additional registers, after some minor refactoring to make things a bit more clear.

Fixes #251.

### CODE REVIEW

File: gematria/datasets/find_accessed_addrs_exegesis.cc [hunk 1]
```diff
@@ -134,21 +135,30 @@ Expected<ExecutionAnnotations> ExegesisAnnotator::findAccessedAddrs(
   BenchCode.Key.MemoryValues["memdef1"] = MemVal;
 
   const llvm::MCRegisterInfo &MRI = State.getRegInfo();
+  std::vector<unsigned> UsedRegisters = gematria::getUsedRegisters(
```
Comment (ondrasej) [inline gematria/datasets/find_accessed_addrs_exegesis.cc:138]:
```cc
const std::vector<unsigned>
```

File: gematria/datasets/find_accessed_addrs_exegesis.cc [hunk 2]
```diff
@@ -211,19 +221,8 @@ Expected<ExecutionAnnotations> ExegesisAnnotator::findAccessedAddrs(
     MemAnnotations.add_accessed_blocks(Mapping.Address);
   }
 
-  std::vector<unsigned> UsedRegisters = gematria::getUsedRegisters(
-      *DisInstructions, State.getRegInfo(), State.getInstrInfo());
-
   MemAnnotations.mutable_initial_registers()->Reserve(UsedRegisters.size());
```
Comment (ondrasej) [inline gematria/datasets/find_accessed_addrs_exegesis.cc:224]: This line should move along with the rest of the code (it's pre-allocating memory for all the `add_initial_regisers()` calls that were in the for loop below.

File: gematria/datasets/find_accessed_addrs_exegesis.cc [hunk 3]
```diff
@@ -134,21 +135,30 @@ Expected<ExecutionAnnotations> ExegesisAnnotator::findAccessedAddrs(
...
+
+    if (MRI.getRegClass(X86::GR64_NOREX2RegClassID).contains(RegisterIndex)) {
+      RegVal.Value = APInt(64, kInitialRegVal);
+      NewRegisterValue->set_register_value(kInitialRegVal);
+    } else if (MRI.getRegClass(X86::VR128RegClassID).contains(RegisterIndex)) {
+      RegVal.Value = APInt(128, kInitialRegVal);
+      NewRegisterValue->set_register_value(kInitialRegVal);
+    } else if (RegisterIndex == X86::EFLAGS) {
+      RegVal.Value = 0;
+      NewRegisterValue->set_register_value(0);
```
Comment (ondrasej) [inline gematria/datasets/find_accessed_addrs_exegesis.cc:157]: Some bits in `EFLAGS` can be set explicitly using separate instructions (e.g. the `DF` flag in question is modified using instructions `STD` and `CLD`). Similar for other flags.

I don't think we need to do anything about it now, but we should keep this in mind, and we might want to set it/allow setting it in the future.

**Review by ondrasej — CHANGES_REQUESTED** [review]

File: gematria/datasets/find_accessed_addrs_exegesis.cc (hunk 3 above)
Comment (boomanaiden154) [reply gematria/datasets/find_accessed_addrs_exegesis.cc:157]: Yep. Support would need to be added on the Exegesis side for that. I can look at doing that at some point.

File: gematria/datasets/find_accessed_addrs_exegesis.cc [hunk 4]
```diff
@@ -134,21 +136,49 @@ Expected<ExecutionAnnotations> ExegesisAnnotator::findAccessedAddrs(
...
-    RegisterValue RegVal;
-    RegVal.Register =
-        MRI.getRegClass(X86::GR64_NOREX2RegClassID).getRegister(i);
-    RegVal.Value = APInt(64, kInitialRegVal);
-    BenchCode.Key.RegisterInitialValues.push_back(RegVal);
-  }
+  for (unsigned RegisterIndex : UsedRegisters) {
+    // Skip all the segment registers other than FS and GS as we cannot set
+    // them from userspace and the machine code verifier does not complain if
+    // they are not defined.
```
Comment (ondrasej) [inline gematria/datasets/find_accessed_addrs_exegesis.cc:147]: Is there a register class covering the segment registers? There might even be one covering the user-assignable registers `FS` and `GS` or the hard-coded ones. This is an architecture-defined thing (not LLVM-defined or ABI-defined) so there might very likely be one.

**Review by ondrasej — APPROVED** [review]

File: gematria/datasets/find_accessed_addrs_exegesis.cc (hunk 4 above)
Comment (boomanaiden154) [reply gematria/datasets/find_accessed_addrs_exegesis.cc:147]: There's one for the segment registers, but not one that explicitly excludes FS/GS.

**Pull request merged**